cpp/src/reductions/reduction_operators.cuh

```diff
@@ -0,0 +1,315 @@
+        T_int assumed ;
+
+        do {
+            assumed = old;
+            T target_value = (is_32_align) ? T(old >> 16) : T(old & 0xffff);
+            T new_value = op(update_value, target_value);
+
+            old = (is_32_align) ? (old & 0xffff) | (new_value << 16)
+                                : (old & 0xffff0000) | new_value;
+            old = atomicCAS(address_uint32, assumed, old);
+        } while (assumed != old);
+    }
+};
+
+// 4 bytes atomic operation
+template<typename T, typename Op>
+struct genericAtomicOperationImpl<T, Op, 4> {
+    __forceinline__  __device__
+    void operator()(T* addr, T const & update_value, Op op)
+    {
+        using T_int = unsigned int;
+
+        T old_value = *addr;
+        T assumed {old_value};
+
+        do {
+            assumed  = old_value;
+            const T new_value = op(update_value, old_value);
+
+            T_int ret = atomicCAS(
+                reinterpret_cast<T_int*>(addr),
+                type_reinterpret<T_int, T>(assumed),
+                type_reinterpret<T_int, T>(new_value));
+
+            old_value = type_reinterpret<T, T_int>(ret);
+
+        } while (assumed != old_value);
+    }
+};
+
+// 8 bytes atomic operation
+template<typename T, typename Op>
+struct genericAtomicOperationImpl<T, Op, 8> {
+    __forceinline__  __device__
+    void operator()(T* addr, T const & update_value, Op op)
+    {
+        using T_int = unsigned long long int;
+
+        T old_value = *addr;
+        T assumed {old_value};
+
+        do {
+            assumed  = old_value;
+            const T new_value = op(update_value, old_value);
+
+            T_int ret = atomicCAS(
+                reinterpret_cast<T_int*>(addr),
+                type_reinterpret<T_int, T>(assumed),
+                type_reinterpret<T_int, T>(new_value));
+            old_value = type_reinterpret<T, T_int>(ret);
+
+        } while (assumed != old_value);
+    }
+};
+
+// intermediate functoin resolve underlying data type
+template <typename T, typename Op>
+__forceinline__  __device__
+void genericAtomicOperationUnderlyingType(
+    T* addr, T const & update_value, Op op)
+{
+    genericAtomicOperationImpl<T, Op, sizeof(T)>()
+        (addr, update_value, op);
+}
+
+// The generic atomic operation with binary operator
+template <typename T, typename Op>
+__forceinline__  __device__
+void genericAtomicOperation(T& existing_value, T const & update_value, Op op)
+{
+    // type cast to underlying data type for wrapper types
+    genericAtomicOperationUnderlyingType(
+        & cudf::detail::unwrap(existing_value),
+        cudf::detail::unwrap(update_value),
+        op
+    );
+}
+
+// ------------------------------------------------------------------------
+// specialized functions for operators
+// `atomicAdd` supports int32, float, double (signed int64 is not supproted.)
+// `atomicMin`, `atomicMax` support int32_t, int64_t
+
+template<>
+struct genericAtomicOperationImpl<float, DeviceSum, 4> {
```

Comment:
I assume these won't duplicate the ones added in the atomics PR?

Reply:
Will remove since atomic PR is meged